File: semant_demo_backend/semant_demo/schema/documents.py
from pydantic import BaseModel
from uuid import UUID
from datetime import datetime

class Document(BaseModel):
    id: UUID

    title: str | None = None
    public: bool | None = None
    documentType: str | None = None
    partNumber: str | None = None
    dateIssued: datetime | None = None
    yearIssued: int | None = None
    language: str | None = None
    publisher: str | None = None
    placeOfPublication: str | None = None
    subtitle: str | None = None
    editors: list[str] | None = None
    partName: str | None = None
    seriesName: str | None = None
    edition: str | None = None
    author: list[str] | None = None
    illustrators: list[str] | None = None
    translators: list[str] | None = None
    redaktors: list[str] | None = None
    seriesNumber: str | None = None
    keywords: list[str] | None = None
    
class DocumentBrowse(BaseModel):
    items: list[Document]
    next_offset: int | None = None
    has_more: bool
    total_count: int

class DocumentStats(BaseModel):
    document_id: str
    collection_id: str
    chunks_in_collection: int
    total_chunks: int
    annotations_count: int
    distinct_tags_count: int

File: semant_demo_frontend/src/schemas/documents.ts
import { z } from 'zod'

const DocumentSchema = z.object({
  id: z.string(),

  title: z.string().optional(),
  public: z.boolean().optional(),
  documentType: z.string().optional(),
  partNumber: z.string().optional(),
  dateIssued: z.iso.datetime().optional().transform((date) => date ? new Date(date) : undefined),
  yearIssued: z.number().optional(),
  language: z.string().optional(),
  publisher: z.string().optional(),
  placeOfPublication: z.string().optional(),
  subtitle: z.string().optional(),
  editors: z.array(z.string()).optional(),
  partName: z.string().optional(),
  seriesName: z.string().optional(),
  edition: z.string().optional(),
  author: z.array(z.string()).optional(),
  illustrators: z.array(z.string()).optional(),
  translators: z.array(z.string()).optional(),
  redaktors: z.array(z.string()).optional(),
  series_number: z.string().optional(),
  keywords: z.array(z.string()).optional()
})

const DocumentsSchema = z.array(DocumentSchema)

const DocumentBrowseParamsSchema = z.object({
  collectionId: z.uuid().optional(),
  limit: z.number().int().positive(),
  offset: z.number().int().min(0),
  sortBy: z.string().optional(),
  sortDesc: z.boolean().optional(),
  title: z.string().optional(),
  author: z.string().optional(),
  publisher: z.string().optional(),
  documentType: z.string().optional()
})

const DocumentBrowseResponseSchema = z.object({
  items: DocumentsSchema,
  nextOffset: z.number().int().nullable().optional(),
  hasMore: z.boolean(),
  totalCount: z.number().int().min(0)
})

export {
  DocumentSchema,
  DocumentsSchema,
  DocumentBrowseParamsSchema,
  DocumentBrowseResponseSchema
}
